rustls/src/lib.rs

```diff
@@ -638,6 +638,41 @@ pub mod manual;
 
 pub mod time_provider;
 
+/// APIs for implementing locks on no-std environments.
+pub mod lock {
+    #[cfg(not(feature = "std"))]
+    use alloc::boxed::Box;
+
+    #[cfg(not(feature = "std"))]
+    /// A lock protecting shared data.
+    pub trait Lock: Send + Sync {
+        /// The state protected by the lock.
+        type Data: Send;
+
+        /// Acquire the lock.
+        fn lock(&self) -> MutexGuard<'_, Self::Data>;
+    }
+
+    #[cfg(not(feature = "std"))]
+    /// A lock builder.
+    pub trait MakeMutex {
```

Comment:
So the setup here seems to try to mimic the std API as much as possible to minimize the amount of code to change in the library. I think on balance I'd prefer to go the other way to make it more obvious that this isn't exactly the default stuff, and use `Arc<dyn Lock<T>>` explicitly at call sites.

It would be nice to fold the `MakeMutex` trait into a `Lock::new()` method if possible.

(I guess this assumes that the intermediate vtable does not cause substantial performance degradation for the case where `std` is available. I think that holds because we don't use `Mutex`es all that much and not never inside particularly hot code?)

Reply:
> So the setup here seems to try to mimic the std API as much as possible to minimize the amount of code to change in the library. I think on balance I'd prefer to go the other way

@ctz I think both approaches here are reasonable, but I'm swayed by djc's position that it should be obvious when we're deviating from the standard path. Do you agree or would you prefer we keep the original approach that minimizes the lib changes?

Reply:
I don't mind the call sites knowing about the difference. I think I would be concerned if `std` users are paying extra runtime cost (specifically in the lock/unlock path; not concerned about mutex construction or memory footprint) for the benefit of this extra abstraction?

Reply:
> (I guess this assumes that the intermediate vtable does not cause substantial performance degradation for the case where std is available. I think that holds because we don't use Mutexes all that much and not never inside particularly hot code?)

Just want to emphasize something I said in the other comment thread; there is no vtable in play for the case where std is available. The `crate::lock::Mutex` type in that case is just a re-export of `std::sync::Mutex`. We only use the `Lock` trait for no-std.

Reply:
> I think on balance I'd prefer to go the other way to make it more obvious that this isn't exactly the default stuff, and use Arc<dyn Lock<T>> explicitly at call sites.

Do you mean dropping the `pub(crate) type Mutex<T> = alloc::sync::Arc<dyn Lock<Data = T>>;` reexport from `crate::lock`` and using `Arc<dyn Lock<Data = T>>` in all of the places that used `crate::lock::Mutex` , or something else?

> It would be nice to fold the MakeMutex trait into a Lock::new() method if possible.

I think that might be exclusive with keeping the `Lock` trait compatible a trait object no?

Reply:
> One thing I missed before is that the Lock here is taking an associated Data type, but I think that should be a type argument on the trait instead of an associated type 

Fixed.

> The other way to go here would be to provide our own Mutex type that wraps around an inner: std::sync::Mutex<T> in the std case or an inner: Arc<dyn Lock> in the no-stdcase.

I implemented this approach. I also lifted the lock module out of `lib.rs` so I could more easily create two side-by-side impls with minimal gating.

> if we use a local Mutex type to abstract over the options I think that type could also have a new() method and this issue would fall away.

The `MakeMutex` trait remains in my latest push. I personally don't see how it can be avoided but would be open to new ideas. 

The type being guarded by the mutex is crate internal in some instances (e.g. `limited_cache::LimitedCache`), so the caller can't construct a `Arc<dyn Lock<T>>` to give to us in `new()` because `T` is private. On the other hand, we can't construct the `Arc<dyn Lock<T>>` ourselves in `new()` without something like `MakeMutex` because we don't know the concrete choice of `dyn Lock` the caller wants to use.

We can probably fiddle with the shape of `MakeMutex` (maybe it could be a `Fn` provided by arg instead of a trait?) but in general I'm getting a little bit weary of refactoring this branch around through many different permutations. Is there a specific downside, or issue with, the `MakeMutex` approach you are trying to avoid?

Reply:
Yup, agree that `MakeMutex` can't be avoided -- sorry for the noise.